src/App.tsx

```diff
@@ -1,61 +1,122 @@
-        </div>
-
-        <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
-          </select>
-
-          <span className="error">Please choose a user</span>
-        </div>
-
-        <button type="submit" data-cy="submitButton">
-          Add
-        </button>
-      </form>
-
-      <section className="TodoList">
-        <article data-id="1" className="TodoInfo TodoInfo--completed">
-          <h2 className="TodoInfo__title">
-            delectus aut autem
-          </h2>
-
-          <a className="UserInfo" href="mailto:Sincere@april.biz">
-            Leanne Graham
-          </a>
-        </article>
-
-        <article data-id="15" className="TodoInfo TodoInfo--completed">
-          <h2 className="TodoInfo__title">delectus aut autem</h2>
-
-          <a className="UserInfo" href="mailto:Sincere@april.biz">
-            Leanne Graham
-          </a>
-        </article>
-
-        <article data-id="2" className="TodoInfo">
-          <h2 className="TodoInfo__title">
-            quis ut nam facilis et officia qui
-          </h2>
-
-          <a className="UserInfo" href="mailto:Julianne.OConner@kory.org">
-            Patricia Lebsack
-          </a>
-        </article>
-      </section>
-    </div>
-  );
-};
+type State = Todo;
+
+export class App extends React.Component<{}, State> {
+  state = {
+    id: Math.max(...todos.map(todo => todo.id)) + 1,
+    title: '',
+    userId: 0,
+    completed: false,
+  };
+
+  inputStatusMonitoring = false; // this property used in order to comply with this requirement: "errors should appear only after clicking the `Add` button"
+
+  handleChange = (event:
+  React.FormEvent<HTMLInputElement>
+  | React.ChangeEvent<HTMLSelectElement>) => { // 2 types for events of input change ang select tag change
+    const { name, value } = event.target as HTMLInputElement;
+
+    if (value.match(/^[A-Za-zА-Яа-я0-9 ]*$/i)) {
+      this.setState((prevState => ({ // used prevState to avoid type script error (2345)
+        ...prevState,
+        [name]: value,
+      })));
+    }
+  };
+
+  handleSubmit = () => {
+    if (this.state.userId && this.state.title.length) {
+      todos.push({
+        ...this.state,
+        user: getUser(this.state.userId),
+      });
+      this.iterateEventId();
+      this.clearTheForm();
+    } else {
+      this.inputStatusMonitoring = true;
+      this.forceUpdate();
+    }
+  };
+
+  iterateEventId = () => {
+    this.setState(state => ({
+      id: state.id + 1,
+    }));
+  };
+
+  clearTheForm = () => {
+    this.setState({
+      title: '',
+      userId: 0,
+    });
+    this.inputStatusMonitoring = false;
+  };
+
+  render() {
+    return (
+      <div className="App">
+        <h1>Add todo form</h1>
+
+        <form
+          action="/api/users"
+          method="POST"
+          onSubmit={(e) => {
+            e.preventDefault();
+            this.handleSubmit();
+          }}
```

Comment:
move prevent into handleSubmit function
```suggestion
          onSubmit={this.handleSubmit}
```

Reply:
I've done it to avoid manual definition of the type of the event, to minimize error chances. So the compact code is the priority?